### ISSUE TITLE
Document protocol / provide gRPC `service` definition

### ISSUE THREAD

**Issue by thomastoye** [jketterl/codecserver#8]
Thanks for starting this project. I want to take it for a spin, but it's unclear how I should invoke it. I'm not clear whether the communication occurs over gRPC or if the protobufs are just used to encode data. I assume the latter based on this line in the `README`:

>It uses a protocol based on protocol buffers (or protobuf for short) to communicate, exchanging length-delimited Any-encapsulated messages on the sockets.

Could you document the protocol? I [see](https://github.com/jketterl/codecserver/blob/develop/src/server/clientconnection.cpp#L61) the following messages are supported to be received: `Request`, `Check`, `ChannelData`, `SpeechData`, `Renegotiation`.

For `[server:tcp]; port=1073; bind=::`, do I assume right that I make a TCP connection to `1073`, and then the protocol should work as follows (for AMBE->raw):

1. Optional: I start by sending a `Check` with `codec=ambe` message, as a sanity check and to make sure we can connect to the server
2. I send a `Request` message, with `Direction.DECODE`. This will start a new session (and only one session can be in progress). I'm unsure why `directions` is repeated - does the order matter or does it just set up `Session`s to process both `ChannelData` and `SpeechData` (when I want to do both decoding and encoding, should I always go `ABABABAB` or can I interleave them as I want like `AAABABBABAA`)?
3. I start sending `SpeechData` and I receive nothing back. The data is sent to the device async. 
4. When I want to read the decoded data, what do I do? I think the data is sent directly on the socket?

I am also unclear on the data format for `SpeechData` and `ChannelPacket` (which I assume is the decoded/unencoded speech data). I guess these will be codec and device-dependent? I think `FramingHint` describes this, is that correct? For the `ambe3k` driver in particular, is the output encoded with PCM3500?

Would you consider using gRPC or adding a `service` definition to the `.proto` files? It would greatly simplify implementing client due to code generation. Or if you answer my questions here, I will gladly help to document the protocol in the README.

**Comment by thomastoye** [comment]
Hi @jketterl, posting here since I have no edit rights to the wiki. Could you proofread the below and add it to the wiki? Thanks.

--

# Protocol

## Communicating with a `codec-server`

### Message wire format

[Protobuf](https://developers.google.com/protocol-buffers/) is used for serialization. The `.proto` files can be found in [`src/lib/proto`](https://github.com/jketterl/codecserver/blob/develop/src/lib/proto).

Messages on the wire are preceded by an unsigned integer in [Protobuf Varint](https://developers.google.com/protocol-buffers/docs/encoding#varints) encoding. This indicates the length of the following message.

The message itself is packed in an [`Any`](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Any) message. This allows recovery of the type. In C++ and Java, `Coded{Input,Output}Stream`s may be used to simplify integration.

To receive a message without such support:

1. Read the `Varint` to know the expected length of the message
1. Read the message as an [`Any`](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Any)
1. Use the `type_url` to construct an object of the appropriate type

To send a message without such support:

1. Pack the message in an [`Any`](https://developers.google.com/protocol-buffers/docs/reference/google.protobuf#google.protobuf.Any) (e.g. `any = google.protobuf.any_pb2.Any(); any.Pack(my_msg)`)
1. Get the size in bytes of the message and encode it as a `Varint`
1. Send this size in bytes, then the message over the wire

### Connecting

Connect to the server, either over TCP or over UNIX domain sockets. Check the `codecserver.conf` (`[server:...]` section) of the server to see what you have enabled for your server and what the port or socket path is.

### Setup

When connecting, the server will send you a `Handshake` message. You can use the `serverVersion` to ensure you have a compatible implementation.

To check if the server has support for the codec you want to use, send a `Check` message. If the codec is supported, you will received a `Response` message with `Status: OK`. You can also immediately try to start a session with a codec and see if that is successful, `Check` is mostly meant as a stand-alone message to see what codecs are supported.

### Starting a session

To start a session, send a `Request` with the `codec` you want to use. You can select the directions you want to use (encoding, decoding, or both), some devices may not implement both (software devices in particular). The codec you want to use may require you to provide `args`, these are codec-specific. For example, the `ambe` codec can take a `rate` or a `ratep` string.

In the `Response`, the server may indicate the expected framing for the codec and argument you selected. This describes the number of bits expected for encoded/decoded samples.

### Coding audio

To encode or decode audio, send `SpeechData` or `ChannelData`, respectively. Coding is an asynchronous process and you should not await this synchronously.

To encode data, send `SpeechData` messages to the server. At some point you will receive `ChannelData` back, this is the encoded data. To decode data, send `ChannelData` messages to the server. At some point you will receive `SpeechData` back, this is the decoded data.

The format of both `SpeechData` and `ChannelData` will be specific to the codec and its arguments.

**Comment by jketterl** [comment]
I just came back to this after some time... I do have to apologize, I didn't really pay much attention to this project once it had reached a good enough state to work for what I needed it for.

I checked out your example and got the same results. Looking at the code, I found that channel packets (that's packets containing encoded speech data from the stick) were not handled at all in one particular spot in the code.

I have now added the missing implementation and I believe your example does now work better, at least it looks like it is actually receiving data, and it is also completing the example file all the way to the end.

Also, on the documentation: Thank you for writing up the protocol documentation. I have added it as a new page on the wiki, I'm going to do the proofreading there in a minute. The wiki is also open to edit, so feel free to add things there if necessary.
